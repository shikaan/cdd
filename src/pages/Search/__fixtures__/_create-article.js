export const createArticle = (topics, title = "title") => {
  return {
    node: {
      slug: "slug",
      title,
      publishDate: "June 14, 2020",
      tags: topics ?? [
        "cache",
        "frontend",
        "backend",
        "python"
      ],
      body: {
        "childMarkdownRemark": {
          "timeToRead": 7
        }
      },
      coverImage: {
        fluid: {
          base64: "data:image/jpeg;base64,/9j/4AAQSkZJRgABAQEAlgCWAAD/2wBDAAoHBwkHBgoJCAkLCwoMDxkQDw4ODx4WFxIZJCAmJSMgIyIoLTkwKCo2KyIjMkQyNjs9QEBAJjBGS0U+Sjk/QD3/2wBDAQsLCw8NDx0QEB09KSMpPT09PT09PT09PT09PT09PT09PT09PT09PT09PT09PT09PT09PT09PT09PT09PT09PT3/wAARCAAKABQDASIAAhEBAxEB/8QAGQAAAgMBAAAAAAAAAAAAAAAAAAcCBAUG/8QAJRAAAgEEAAQHAAAAAAAAAAAAAQIDAAQFEQYhMWETIkFCgbHR/8QAFgEBAQEAAAAAAAAAAAAAAAAABAID/8QAFxEBAQEBAAAAAAAAAAAAAAAAAAECEv/aAAwDAQACEQMRAD8AlZcF44YjVvtZinllcltHuOhFUZuD7lOTyRuoHNktgft64fFZG9AVhd3GweR8Vv2m+0jmwjYu2zGpJ32odJhT3mJx6XLCTLGF/VJLFkI+KK085PKck+5X6D3GitJuo5f/2Q==",
          aspectRatio: 1.965,
          src: "//images.ctfassets.net/rrd69nadniqd/5JjQhwkcdHbYN2kWfzAlib/5ac520012c6b363b334777752963fc1b/cover.jpg?w=800&q=50",
          srcSet: "//images.ctfassets.net/rrd69nadniqd/5JjQhwkcdHbYN2kWfzAlib/5ac520012c6b363b334777752963fc1b/cover.jpg?w=200&h=102&q=50 200w,\n//images.ctfassets.net/rrd69nadniqd/5JjQhwkcdHbYN2kWfzAlib/5ac520012c6b363b334777752963fc1b/cover.jpg?w=400&h=204&q=50 400w,\n//images.ctfassets.net/rrd69nadniqd/5JjQhwkcdHbYN2kWfzAlib/5ac520012c6b363b334777752963fc1b/cover.jpg?w=786&h=400&q=50 786w",
          sizes: "(max-width: 800px) 100vw, 800px"
        }
      }
    }
  };
};
