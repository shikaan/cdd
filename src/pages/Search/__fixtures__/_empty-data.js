export const emptyData = {
  posts: {
    edges: []
  },
  content: {
    title: ""
  }
};
