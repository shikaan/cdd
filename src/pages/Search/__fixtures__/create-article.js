export const createArticle = (topics, title = 'title') => {
  return {
    node: {
      fields: {
        readingTime: {
          minutes: 1
        }
      },
      frontmatter: {
        title,
        date: "date",
        tags: topics,
        coverImage: {
          childImageSharp: {
            fixed: {
              width: 112,
              height: 112,
              base64: "base64",
              tracedSVG: "tracedSVG",
              aspectRatio: 1,
              src: "src",
              srcSet: "srcSet",
              srcWebp: "srcWebp",
              srcSetWebp: "srcSetWebp",
              originalName: "originalName"
            }
          }
        }
      }
    }
  }
}
