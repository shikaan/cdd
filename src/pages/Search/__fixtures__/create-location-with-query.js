export const createLocationWithQueryString = (query) => {
  return {
    search: query ? `?${query}` : ""
  };
};
