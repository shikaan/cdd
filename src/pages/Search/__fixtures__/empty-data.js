export const emptyData = {
  posts: {
    edges: []
  }
}
