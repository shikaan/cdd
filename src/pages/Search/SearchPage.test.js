import React from "react";
import {shallow, mount} from "enzyme";
import "jest-styled-components";

import {createArticle} from "./__fixtures__/create-article";
import {emptyData} from "./__fixtures__/empty-data";
import {createLocationWithQueryString} from "./__fixtures__/create-location-with-query";

import SearchPage from ".";

describe("SearchPage", () => {
  it("renders correctly", () => {
    const wrapper = shallow(
      <SearchPage
        content={{}}
        data={emptyData}
        location={createLocationWithQueryString("query=foo")}
      />);

    expect(wrapper).toMatchSnapshot();
  });

  describe("getSearchQuery", () => {
    it("returns an empty string if no query string", () => {
      const wrapper = mount(
        <SearchPage
          content={{}}
          data={emptyData}
          location={createLocationWithQueryString()}
        />);

      expect(wrapper.instance().getSearchQuery()).toHaveLength(0);
    });

    it("returns an the query under `query`", () => {
      const wrapper = mount(
        <SearchPage
          content={{}}
          data={emptyData}
          location={createLocationWithQueryString("query=foo")}
        />);

      expect(wrapper.instance().getSearchQuery()).toEqual("foo");
    });
  });

  describe("getTrendingTopics", () => {
    let getTrendingTopics;
    beforeAll(() => {
      const wrapper = mount(
        <SearchPage
          content={{}}
          data={emptyData}
          location={createLocationWithQueryString("sdfg=sdfgsdf")}
        />);

      getTrendingTopics = wrapper.instance().getTrendingTopics;
    });

    it("returns empty array if list is missing", () => {
      const result = getTrendingTopics([]);

      expect(result).toHaveLength(0);
    });

    it("returns no duplicates", () => {
      const result = getTrendingTopics([
        createArticle(["foo", "bar"]),
        createArticle(["foo", "baz"]),
      ]);

      expect(result).toEqual(["foo", "bar", "baz"]);
    });

    it("returns a list of maximum length provided", () => {
      const result = getTrendingTopics([
        createArticle(["foo", "bar"]),
        createArticle(["foo", "baz"]),
        createArticle(["goo", "bas"]),
      ], 3);

      expect(result).toEqual(["foo", "bar", "baz"]);
    });
  });

  describe("performSearch", () => {

    it("sets state only if query is longer than 3", (done) => {
      const data = {
        posts: {
          "edges": [
            createArticle(["foo"], "foo")
          ]
        }
      };

      const wrapper = mount(
        <SearchPage
          content={{}}
          data={data}
          location={createLocationWithQueryString()}
        />);

      const instance = wrapper.instance();

      instance
        .setState(
          {searchQuery: "fo"},
          () => {
            setTimeout(() => {
              expect(instance.state.searchResults).toBeNull();
              done();
            }, SearchPage.DEBOUNCE_INTERVAL * 2); // Wait for debounce to pass
          });
    });
    it("returns articles if query is included in title", (done) => {
      const data = {
        posts: {
          "edges": [
            createArticle(["foo"], "bar"),
            createArticle(["foo"], "baz")
          ]
        }
      };

      const wrapper = mount(
        <SearchPage
          content={{}}
          data={data}
          location={createLocationWithQueryString()}
        />);

      const instance = wrapper.instance();

      instance
        .setState(
          {searchQuery: "bar"},
          () => {
            setTimeout(() => {
              expect(instance.state.searchResults).toHaveLength(1);
              expect(instance.state.searchResults[0]).toHaveProperty("node.frontmatter.title", "bar");

              done();
            }, SearchPage.DEBOUNCE_INTERVAL * 2); // Wait for debounce to pass
          });
    });
    it("returns articles if query is included in topics", (done) => {
      const data = {
        posts: {
          "edges": [
            createArticle(["bar"], "foo"),
            createArticle(["baz"], "foo")
          ]
        }
      };

      const wrapper = mount(
        <SearchPage
          content={{}}
          data={data}
          location={createLocationWithQueryString()}
        />);

      const instance = wrapper.instance();

      instance
        .setState(
          {searchQuery: "bar"},
          () => {
            setTimeout(() => {
              expect(instance.state.searchResults).toHaveLength(1);
              expect(instance.state.searchResults[0]).toHaveProperty("node.frontmatter.tags", ["bar"]);

              done();
            }, SearchPage.DEBOUNCE_INTERVAL * 2); // Wait for debounce to pass
          });
    });
  });
});
