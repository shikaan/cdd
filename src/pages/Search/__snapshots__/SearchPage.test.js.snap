// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchPage renders correctly 1`] = `
<SearchTemplate>
  <SEO
    description="Find articles by topic or title"
    keywords={
      Array [
        "search",
      ]
    }
    lang="en"
    meta={Array []}
    slug="/search"
    title="Search"
  />
  <Header />
  <SearchResults
    articles={Array []}
    content={
      Object {
        "emptyState": Object {
          "cta": "Pick a trending topic",
          "paragraph": Array [
            "Our best detectives could not find any result for your query...",
            "Try again or check one of our coolest topics!",
          ],
          "title": "Aw snap! ðŸ˜­",
        },
        "placeholder": "Start Typing",
        "seo": Object {
          "description": "Find articles by topic or title",
          "keywords": Array [
            "search",
          ],
          "title": "Search",
        },
        "shared": Object {
          "disclaimer": Object {
            "actionLabel": "Contribute",
            "dismissLabel": "Dismiss",
            "text": "This blog is meant to target mobile devices, but contributions for a desktop version are welcomed",
          },
          "readingTime": "min read",
        },
        "subTitle": "Trending Topics",
      }
    }
    performSearch={[Function]}
    searchQuery="foo"
    setSearchQuery={[Function]}
  />
  <Results
    content={
      Object {
        "emptyState": Object {
          "cta": "Pick a trending topic",
          "paragraph": Array [
            "Our best detectives could not find any result for your query...",
            "Try again or check one of our coolest topics!",
          ],
          "title": "Aw snap! ðŸ˜­",
        },
        "placeholder": "Start Typing",
        "seo": Object {
          "description": "Find articles by topic or title",
          "keywords": Array [
            "search",
          ],
          "title": "Search",
        },
        "shared": Object {
          "disclaimer": Object {
            "actionLabel": "Contribute",
            "dismissLabel": "Dismiss",
            "text": "This blog is meant to target mobile devices, but contributions for a desktop version are welcomed",
          },
          "readingTime": "min read",
        },
        "subTitle": "Trending Topics",
      }
    }
    searchResults={null}
    trendingTopics={Array []}
  />
</SearchTemplate>
`;
